src/odemis/gui/comp/viewport.py

```diff
@@ -854,84 +860,124 @@ def setView(self, view, tab_data):
+                else:
+                    proj = projs[0]
         else:
-            stream = streams[0]
+            proj = projs[0]
 
-        if self.stream is stream:
-            logging.debug("Not reconnecting to stream as it's already connected")
+        if self._projection is proj:
+            # logging.debug("not reconnecting to stream as it's already connected")
             return
 
         # Disconnect from the old VAs
-        if self.stream:
-            if hasattr(self.stream, 'polarization'):
-                logging.debug("Disconnecting %s from polarization VA", stream)
-                self.stream.polarization.unsubscribe(self.on_va_pol_change)
-
-        # Connect to the new VA
-        self.stream = stream
-
-        # get polarization positions and set callback function
-        if hasattr(self.stream, "polarization"):
-            logging.debug("Connecting %s to polarization VA", stream)
-            self.update_legend_pol_choices()
-            self.stream.polarization.subscribe(self.on_va_pol_change)
-            self.bottom_legend.Show(True)
-        # if no polarization VA or if no stream (= None)
+        if self._projection:
+            logging.debug("Disconnecting %s from ARRawViewport", self._projection)
+            if hasattr(self._projection, 'polarization'):
+                logging.debug("Disconnecting %s from polarization VA", proj)
+                self._stream.polarization.unsubscribe(self.on_va_change)
+            if hasattr(self._projection, 'polarimetry'):
+                logging.debug("Disconnecting %s from polarimetry VA", proj)
+                self._stream.polarimetry.unsubscribe(self.on_va_change)
+
+        # Connect the new stream
+        self._stream, self._projection = get_original_stream(proj), proj
+        if proj:
+            logging.debug("Connecting %s to ARRawViewport", proj)
+            # get polarization/polarimetry positions and set callback function
+            # viewport associated to the projection should have only polarization or polarimetry VA
+            if hasattr(proj, "polarization"):
+                logging.debug("Connecting %s to polarization VA", proj)
+                name = "Polarization"
+                choices = self._stream.polarization.choices
+                if choices == frozenset(POL_POSITIONS):
+                    choices = POL_POSITIONS  # use the ordered tuple to keep legend ordered later
+                cur_value = self._stream.polarization.value
+                self.update_legend_choices(choices, cur_value, name)
+                self._projection.polarization.subscribe(self.on_va_change)
+                self.bottom_legend.Show(True)
+            elif hasattr(proj, "polarimetry"):
+                logging.debug("Connecting %s to polarimetry VA", proj)
+                name = "Polarimetry"
+                cur_value = self._stream.polarimetry.value
+                choices = collections.OrderedDict()
```

Comment:
The most direct way to write it is:
```
choices = collections.OrderedDict(
   ("aaaa", (MD_POL_DS0...)),
   ("bbbb", (MD_POL_S0...)),
)
```

Reply:
Good idea. I think, there need to be some additioinal brackets. Otherwise it does not work.
```
choices = collections.OrderedDict((
   ("aaaa", (MD_POL_DS0...)),
   ("bbbb", (MD_POL_S0...)),
))
```